libcxx/include/__tree

```diff
@@ -582,10 +585,37 @@ class _LIBCPP_STANDALONE_DEBUG __tree_node : public __tree_node_base<_VoidPtr> {
 public:
   using __node_value_type _LIBCPP_NODEBUG = __get_node_value_type_t<_Tp>;
 
-  __node_value_type __value_;
+// We use a union to avoid initialization during member initialization, which allows us
+// to use the allocator from the container to allocate the node itself
+#ifndef _LIBCPP_CXX03_LANG
 
+private:
+  union {
+    __node_value_type __value_;
+  };
+
+public:
   _LIBCPP_HIDE_FROM_ABI __node_value_type& __get_value() { return __value_; }
+#else
+
+private:
+  _ALIGNAS_TYPE(__node_value_type) unsigned char __buffer_[sizeof(__node_value_type)];
+
+public:
+  _LIBCPP_HIDE_FROM_ABI __node_value_type& __get_value() {
+    return *std::__launder(reinterpret_cast<__node_value_type*>(&__buffer_));
+  }
+#endif
 
+  template <class _Alloc, class... _Args>
+  _LIBCPP_HIDE_FROM_ABI explicit __tree_node(_Alloc& __na, _Args&&... __args) {
+#ifndef _LIBCPP_CXX03_LANG
+    allocator_traits<_Alloc>::construct(__na, std::addressof(__value_), std::forward<_Args>(__args)...);
+#else
+    allocator_traits<_Alloc>::construct(
+        __na, reinterpret_cast<__node_value_type*>(std::addressof(__buffer_)), std::forward<_Args>(__args)...);
```

Comment:
Q: Is this right? or should I be using 
```suggestion
        __na, reinterpret_cast<__node_value_type*>(std::addressof(__buffer_[0])), std::forward<_Args>(__args)...);
```
to be strictly standards compliant?

Reply:
It's clearer to use `reinterpret_cast<__node_value_type*>(__buffer_)`. When we know overloaded `operator&` can't be involved via ADL or class member functions, it's better to use plain `&`.